python/lsst/daf/butler/script/queryDatasets.py

```diff
@@ -0,0 +1,184 @@
+        return [table.getAstropyTable() for table in self.tables.values()]
+
+
+def queryDatasets(repo, glob, collections, where, find_first, show_uri):
+    """Get dataset refs from a repository.
+
+    Parameters
+    ----------
+    repo : `str`
+        URI to the location of the repo or URI to a config file describing the
+        repo and its location.
+    glob : iterable [`str`]
+        A list of glob-style search string that fully or partially identify
+        the dataset type names to search for.
+    collections : iterable [`str`]
+        A list of glob-style search string that fully or partially identify
+        the collections to search for.
+    where : `str`
+        A string expression similar to a SQL WHERE clause.  May involve any
+        column of a dimension table or (as a shortcut for the primary key
+        column of a dimension table) dimension name.
+    find_first : `bool`
+        For each result data ID, only yield one DatasetRef of each DatasetType,
+        from the first collection in which a dataset of that dataset type
+        appears (according to the order of `collections` passed in).  If used,
+        `collections` must specify at least one expression and must not contain
+        wildcards.
+    show_uri : `bool`
+        If True, include the dataset URI in the output.
+    Returns
+    -------
+    datasetTables : `list` [``astropy.table._Table``]
+        A list of astropy tables, one for each dataset type.
+    """
+    butler = Butler(repo)
+
+    dataset = globToRegex(glob)
+    if not dataset:
+        dataset = ...
+
+    if collections and not find_first:
+        collections = globToRegex(collections)
+    elif not collections:
+        collections = ...
+
+    datasets = butler.registry.queryDatasets(datasetType=dataset,
+                                             collections=collections,
+                                             where=where,
+                                             deduplicate=find_first)
+
+    tables = _Tables()
+
+    def addDatasetComponent(datasetRef, name, uri):
+        """Add an entry in the correct output dataset table for the given
+        DatasetRef and component name, and component URI."""
+        datasetTypeName = datasetRef.datasetType.name
+        if name:
+            datasetTypeName = ".".join((datasetTypeName, name))
+        table = tables.getTable(datasetTypeName)
+        if not table:
+            columnNames = ["type", "run", "id", *[str(item) for item in datasetRef.dataId.keys()]]
+            if show_uri:
+                columnNames.append("URI")
+            table = tables.makeTable(datasetTypeName, columnNames)
+        extraValues = [str(value) for value in datasetRef.dataId.values()]
+        if show_uri:
+            extraValues += [uri]
+        table.addRow((datasetTypeName,
+                      datasetRef.run,
+                      datasetRef.id,
+                      *extraValues))
+
+    for datasetRef in datasets:
```

Comment:
In DM-26324 (#396) I am adding support for sorting DatasetRefs. This would really help with reproducibility of the output here since at the moment we are at the mercy of the order returned by the database and that order is going to depend on the database. For example you end up with the detectors not coming out as you would expect so it's hard to work out if something obvious is missing.  I think this means you can group all the returned DatasetRef from queryDatasets into a dict of sets (keyed by DatasetType name) and then sort the DatasetRef out of the sets.